clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -721,12 +730,17 @@ static void VisitAccessorWrapper(CXXRecordDecl *Owner, ParentTy &Parent,
 // A visitor function that dispatches to functions as defined in
 // SyclKernelFieldHandler for the purposes of kernel generation.
 template <typename... Handlers>
-static void VisitRecordFields(RecordDecl::field_range Fields,
+static void VisitRecordFields(CXXRecordDecl *KernelObject,
                               Handlers &... handlers) {
+
+  QualType KernelType = QualType(KernelObject->getTypeForDecl(), 0);
```

Comment:
Are these only allowed to be top-level things?  Can you clarify how this feature is supposed to work? I don't get it yet.

Reply:
We're adding the kernel object as a parameter of the kernel function (just like we did with the fields). Currently in the kernel body, a kernel object clone is generated and initialized using the fields of kernel function. The clone is then subsequently used in kernel body code. We're now exploring the possibility of getting rid of the clone entirely and using this kernel object